feat($compile): allow using bindToController as object, support both new/isolate scopes

caitp opened the pull request:
bindToController is now able to be specified as a convenient object notation:

```
bindToController: {
  text: '@text',
  obj: '=obj',
  expr: '&expr'
},
scope: {
  notBoundToCtrl: '@someOtherAttr'
}
```

It can also be used in conjunction with new scopes, rather than exclusively isolate scopes:

```
bindToController: {
  text: '@text',
  obj: '=obj',
  expr: '&expr'
},
scope: true
```

PING @petebacondarwin / @johnlindquist

I am not super keen on this, because it makes things much crazier than they used to be. That said:
- Currently, `bindToController` does not automatically imply an isolate scope --- maybe it should (I hate this though)
- It could probably do with more errors being thrown when people use it wrong
- The errors it does throw could be better
- The errors it throws are potentially breaking changes :(

Closes #10420

caitp committed:
feat($compile): allow using bindToController as object, support both new/isolate scopes

bindToController is now able to be specified as a convenient object notation:

```
bindToController: {
  text: '@text',
  obj: '=obj',
  expr: '&expr'
},
scope: {}
```

It can also be used in conjunction with new scopes, rather than exclusively isolate scopes:

```
bindToController: {
  text: '@text',
  obj: '=obj',
  expr: '&expr'
},
scope: true
```

Closes #10420

caitp committed:
fix($compile): respect return value from controller constructor

The return value of the controller constructor is now respected in all cases.

If controllerAs is used, the controller will be re-bound to scope. If bindToController is used,
the previous binding $watches (if any) will be unwatched, and bindings re-installed on the new
controller.

petebacondarwin commented on docs/content/error/$compile/noctrl.ngdoc:
```diff
@@ -0,0 +1,68 @@
+@ngdoc error
+@name $compile:noctrl
+@fullName Controller is required.
```
Should this full name be more like: `"controller as" identifier required to bind to controller`?

gkalpak replied:
If I am not mitaken, this error will also be thrown if there is `bindToController: true` but no `controler:...`. So the literals have to account for all cases (i.e. `bindToController: true` but no `controller` or no controller identifier.

petebacondarwin replied:
Yep, I see that now. I was thrown by the main body of the error document, which doesn't really mention that aspect of the error at all.
I wonder if we ought to have two separate errors?

petebacondarwin commented on docs/content/error/$compile/noctrl.ngdoc:
```diff
@@ -0,0 +1,68 @@
+@description
+
+When using the `bindToController` feature of AngularJS, a directive is required
+to have a Controller, in addition to a controller identifier.
```
> When using the `bindToController` feature of AngularJS, a directive is required
to provide the `controller as` identifier, in addition to providing a controller.

gkalpak replied:
Same as above.
<sub>
BTW, I like _"controller identifier"_ (or _"controller alias"_) better than _"`controller as` identifier"_, as the latter might be misunderstood as implying that a `controllerAs` property is mandatory (which is not the case).
Not a big issue though, since the examples below make it clear.
</sub>

petebacondarwin replied:
I like `controller alias`

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -738,6 +740,37 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      var controllerAs = directive.controllerAs;
+      if (!directive.controller || !identifierForController(controller, controllerAs)) {
+        throw $compileMinErr('noctrl',
+              "Cannot bind to controller without directive '{0}'s controller.",
```
"Cannot bind to controller without directive '{0}'s controller and `controller as` identifier."

petebacondarwin replied:
I think we probably need two different errors here. One for a missing controller and one for a missing controller alias.

petebacondarwin commented on test/ng/compileSpec.js:
```diff
@@ -3798,6 +3798,180 @@ describe('$compile', function() {
+    });
+
+
+    it('should put controller in scope when labelled but not using controllerAs', function() {
```
I don't follow the description of this test. The code below does indeed use `controllerAs`.

caitp replied:
It's not using the `controllerAs` property of the DDO.

petebacondarwin replied:
Ah right. I didn't get that "labelled" meant something and that `controllerAs` was referring to the DDO property rather than implying that the controller had an alias.

I am still not sure what is being tested here. Is it that having a `bindToController` property should not prevent the controller from being attached to the scope?

caitp replied:
`bindToController` is actually irrelevant here, but previously the controller label in the string was ignored. It's just simpler to support the controller-as label too, so it's testing that it works.

caitp replied:
although I'm not super keen on it because it creates 2 ways to do a thing, but it can be worried about later

petebacondarwin replied:
OK, we need to change `labelled` to `has a controller alias` or something like that for consistency. It is only now that I realise that `label` has been used interchangeably with `alias` and `identity` throughout the source.

petebacondarwin commented on src/ng/controller.js:
```diff
@@ -1,5 +1,15 @@
+    var match = CNTRL_REG.exec(controller);
+    if (match) return match[3];
+  }
+}
```
Since this function is only used in compile.js perhaps it should be there and we just allow the CNTRL_REG to be a global instead?

caitp replied:
I think it's better to keep it here, and it's probably better to use it in controller.js too, just to make sure the same logic is used in both places

petebacondarwin replied:
In that case this function needs refactoring to return the controller name as well, since that is what is needed in $controller.

All the references to `label`, `ident` and `identifier` in this file ought to be changed to `alias` for consistency.  If you don't like `alias` then you can choose an alternate and change all the references to `ident`, `identifier`, `alias` and `label` to a single consistent term throughout the codebase.

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -722,9 +722,11 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-            directiveName, scopeName, definition);
+            directiveName, scopeName, definition,
+            (isController ? "controller bindings definition" :
+            "isolate scope definition"));
```
nice

petebacondarwin commented:
Apart from splitting the error message, this looks good to me. @gkalpak do you have any further comments.

gkalpak commented:
:+1: for separate error messages.
<sub>
In the original PR, @caitp mentioned something about _potential leaks and performance regressions_.
I don't think it would be much of an issue (but I wouldn't know better); do you think it would make sense to run some benchmarks (if not already done so) ?
</sub>
Other than that it LGTM.

caitp commented:
if you want to release this today, i'll split up the error message and do whatever other cleanup in about 30 minutes.

caitp committed:
docs($compile): create new error for "missing controller identifier"

caitp commented:
updated to have 2 different errors

gkalpak commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+
+When using the `bindToController` feature of AngularJS, a directive is required
+to have a Controller identifier. This can be supplied using the "controllerAs"
+property of the directive object, or alternatively by adding " as ALIAS" to the
```
directive _definition_ object

gkalpak commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+to have a Controller identifier. This can be supplied using the "controllerAs"
+property of the directive object, or alternatively by adding " as ALIAS" to the
+controller name (in either the "controller" property, or where the controller
+was registered).
```
> or where the controller was registered

(With the risk of embarrassing myself :confused:) what do you mean by that ?

petebacondarwin replied:
I am not sure. I don't think you can specify `controllerName as alias` when defining a controller. The only other time would be inside the `ngController` directive, where the controller name is `'@'` and so we get the controller from the `ng-controller` attribute value, which obviously can also have the `controllerName as alias`.

This phrase should either be clarified or removed.

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+@description
+
+When using the `bindToController` feature of AngularJS, a directive is required
+to have a Controller identifier. This can be supplied using the "controllerAs"
```
Let's standardize on always calling this identifier an **alias** throughout the code and the docs.

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+@ngdoc error
+@name $compile:noident
```
`$compile:noalias`

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
````diff
@@ -0,0 +1,71 @@
+For example, the following directives are valid:
+
+```js
+// OKAY, because controller is a string with a label component.
````
```js
// OKAY, because controller is a string, with an alias component.
```

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+
+
+// OKAY, because the directive uses the controllerAs property to override
+// the controller identifier.
```
```js
// the controller alias
```

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
````diff
@@ -0,0 +1,71 @@
+While the following are invalid:
+
+```js
+// BAD, because the controller property is a string with no identifier.
````
```js
// BAD, because the controller property is a string with no alias.
```

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+directive("bad", function() {
+  return {
+    bindToController: true,
+    controller: "unlabeledCtrl",
```
```js
controller: "unaliasedCtrl",
```

petebacondarwin commented on docs/content/error/$compile/noident.ngdoc:
```diff
@@ -0,0 +1,71 @@
+});
+
+
+// BAD because the controller is not a string (therefore has no identifier),
```
```js
// BAD because the controller is not a string (therefore has no alias),
```

petebacondarwin commented on src/.jshintrc:
```diff
@@ -153,6 +153,9 @@
     "urlIsSameOrigin": false,
 
+    /* ng/controller.js */
+    "identifierForController": false,
```
```js
"aliasForController": false,
```

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -738,6 +740,43 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    }
+    if (isObject(bindings.bindToController)) {
+      var controller = directive.controller;
+      var controllerAs = directive.controllerAs;
```
```js
var controllerAlias = directive.controllerAs;
```

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -738,6 +740,43 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+        throw $compileMinErr('noctrl',
+              "Cannot bind to controller without directive '{0}'s controller.",
+              directiveName);
+      } else if (!identifierForController(controller, controllerAs)) {
```
```js
 } else if (!identifierForController(controller, controllerAlias)) {
```

gkalpak replied:
Or even go one step further and rename `identifierForController` to `aliasForController` :P

petebacondarwin replied:
:-P - quite right! I knew I would miss one...

petebacondarwin commented on src/ng/controller.js:
```diff
@@ -1,5 +1,15 @@
 'use strict';
 
+var CNTRL_REG = /^(\S+)(\s+as\s+(\w+))?$/;
+function identifierForController(controller, ident) {
```
`aliasForController`

petebacondarwin commented:
@caitp  - can you take a look at the comments here and update this PR? Thanks

caitp commented:
I don't think we should be calling it "alias", because it's misleading --- so I'm just going to ignore all of the comments about renaming "identifier" to "alias"

petebacondarwin commented:
@caitp - it is fine for you not to think we should call it alias. As I explained in [this comment](https://github.com/angular/angular.js/pull/10467/files#r23216369), what we do need is consistency.

I assume that your preferred name is `identifier`. In which case please go through and make sure that this is used throughout. In particular remove any use of `alias` and `label`.

frfancha commented:
@caitp 
You say "the important thing is that people generally like controllerAs and want to use it".
I wonder why. In our experiences directives with controller are mush slower than directives with just link function and scope access, especially used under ng-repeat

caitp committed:
docs($compile,$route): reword "controller alias" to clarify

Clarify that these aliases are identifier names used to reference the controller.

caitp commented:
this is ready to merge

petebacondarwin commented:
Great! LGTM! Thanks @caitp. I will merge tomorrow unless you want to do it today.

caitp commented:
I've got it, thanks for reviewing
